A75: xDS Aggregate Cluster Behavior Fixes (#405)
grpc/proposal: A75-xds-aggregate-cluster-behavior-fixes.md

A75: xDS Aggregate Cluster Behavior Fixes
----
* Author(s): @markdroth
* Approver: @ejona86, @dfawley
* Status: {Draft, In Review, Ready for Implementation, Implemented}

> * Implemented in: <language, ...>

Reviewer: ```suggestion
* Implemented in: cpp since v1.61.0
```

Author: Done.

priorities.

### Related Proposals:
* [A37: xDS Aggregate and Logical DNS Clusters][A37]
* [A50: gRPC xDS Outlier Detection Support][A50]
* [A55: xDS-Based Stateful Session Affinity for Proxyless gRPC][A55]
* [A60: xDS-Based Stateful Session Affinity for Weighted Clusters][A60]
* [A56: Priority LB policy][A56]
* [A61: IPv4 and IPv6 Dualstack Backend Support][A61]
* [A74: xDS Config Tears (pending)][A74]

[A37]: A37-xds-aggregate-and-logical-dns-clusters.md
[A50]: A50-xds-outlier-detection.md
[A55]: A55-xds-stateful-session-affinity.md
[A56]: A56-priority-lb-policy.md
[A60]: A60-xds-stateful-session-affinity-weighted-clusters.md
[A61]: A61-IPv4-IPv6-dualstack-backends.md
[A74]: https://github.com/grpc/proposal/pull/404

## Proposal

There are two main parts to this proposal: aggregate cluster changes and
stateful session affinity changes.

### Aggregate Cluster Changes

Instead of representing an aggregate cluster as a single cluster whose
priorities come from different underlying clusters, we will instead
represent an aggregate cluster as an instance of a priority LB policy
([gRFC A56][A56]) where each child is a cds LB policy for the underlying

> cluster.

Reviewer 2: If an aggregate cluster has a child that is also an aggregate cluster, should that be represented as a deep tree, with each child of the root priority LB policy corresponding to a direct child of the root aggregate cluster, or a flat tree, with each child of the root priority LB policy corresponding to an underlying cluster?

Author: The intent is that the code in the `XdsDependencyManager` will recursively resolve the aggregate cluster dependency tree, using the same algorithm that we'd already implemented in the cds LB policy prior to this design.  In the complete `XdsConfig` that is passed down to the LB policies, the data for an aggregate cluster will contain the ordered, flattened list of leaf policies, ignoring any intermediate aggregate clusters.

In other words, let's say you have the following clusters:

A: aggregate cluster pointing to [B, C]
B: EDS cluster
C: aggregate cluster pointing to [D, E]
D: EDS cluster
E: EDS cluster

The `XdsConfig` data for A will indicate the leaf clusters as [B, D, E].  It will not include cluster C.

There's a note about this in the "Rationale" section.

Reviewer 2: Got it. I didn't expect that information to only be in the Rationale section.

So, does this mean that all the root cds LB policy does is read the list of underlying cds LB configs and generate a priority LB policy with one of those configs in each priority?

Author: Yes, the cds policy for the aggregate cluster will generate a config for a priority policy where each child is a cds policy for the underlying leaf cluster.  However, note that the LB policy configs for the cds policies for the underlying clusters do not need to include anything other than the name of the underlying cluster, because the rest of the info will come from the relevant entry in the `XdsConfig` that is passed down via an attribute.

The fact that the cds policy receives just the list of leaf clusters is actually covered in A74, not in this gRFC.  All we're doing here is changing the way we're handling that list in the cds policy.

  is not present, then SSA is not being used for the request, and we do not
  need to retain ownership of the subchannel.
- The xds_override_host policy will have a timer that runs periodically
  to do a sweep over the subchannel map and unref any owned subchannel
  whose `last_used_time` is older than the `idle_timeout` value from the
  CDS resource.  The timer can initially be set for `idle_timeout`
  duration, and subsequent runs can dynamically determine when to run
  based on the `last_used_time` of the entries in the map, with a
  minimum of 5 seconds between runs to minimize overhead.
- When the child policy unrefs a subchannel it owns, if the entry's
  `last_used_time` is newer than the `idle_timeout`, the
  xds_override_host policy assumes ownership of the subchannel instead
  of deleting it.
- In the picker, when we iterate through the addresses in the cookie, if
  we do not find any entries with a subchannel is states READY, IDLE, or
  CONNECTING and we do find entries that do not have any subchannel, we
  will trigger creation of a subchannel for one of those entries.

The picker logic will now look like this:

> ```

Reviewer: nit
```suggestion
```py
```

Author: This is not actually intended to be valid python code; it's just python-style pseudo-code.  I've add a parenthetical indicating this.

Reviewer: Up to you - even if it's not python, it'll still be easier to ready with syntax highlighting. It's clear now from your clarification it's pseudo-code, IMO adding `py` just for readability might worth it.

def Pick(pick_args):
  override_host_attribute = pick_args.call_attributes.get(attribute_key)
  if override_host_attribute is not None:
    entry_with_no_subchannel = None
    idle_subchannel = None
    found_connecting = False
    for address in override_host_attribute.cookie_address_list:
      entry = lb_policy.address_map[address]

>       if (entry found AND
>           entry.health_status is in policy_config.override_host_status):

Reviewer: ```suggestion
     if entry and entry.health_status in policy_config.override_host_status:
```

1. `if entry` implies it's found
2. lowercase and
3. `is in` -> `in`

>         if entry.subchannel is set:

Reviewer: ```suggestion
        if entry.subchannel is not None:
```

>           if entry.subchannel.connectivity_state == READY:
>             override_host_attribute.set_actual_address_list(entry.address_list)
>             return entry.subchannel as pick result
>           elif entry.subchannel.connectivity_state == IDLE:
>             if idle_subchannel is None:
>               idle_subchannel = entry.subchannel
>           elif entry.subchannel.connectivity_state == CONNECTING:
>             found_connecting = True

Reviewer: ```suggestion
                    match entry.subchannel.connectivity_state:
            case READY:
                override_host_attribute.set_actual_address_list(entry.address_list)
                return entry.subchannel   # as pick result
            case IDLE:
              if idle_subchannel is None:
                idle_subchannel = entry.subchannel
            case CONNECTING:
                found_connecting = True
```

we finally have switch/case now :)

Reviewer: Python or not, switch-case just seems more readable to me.

>         else if entry_with_no_subchannel is None:

Reviewer: ```suggestion
        elif entry_with_no_subchannel is None:
```

Reviewer: minor note: `elif` is used everywhere else in this code

      return queue as pick result
    # No READY or IDLE subchannels.  If we found a CONNECTING
    # subchannel, queue the pick and wait for the connection attempt
    # to complete.
    if found_connecting:
      return queue as pick result
    # No READY, IDLE, or CONNECTING subchannels.  If we found an entry
    # with no subchannel, queue the pick and create a subchannel for
    # that entry.
    if entry_with_no_subchannel is not None:
      hop into control plane to create subchannel for entry_with_no_subchannel
      return queue as pick result
  # pick_args.override_addresses not set or did not find a matching subchannel,
  # so delegate to the child picker.
  result = child_picker.Pick(pick_args)
  if result.type == PICK_COMPLETE:
    entry = lb_policy.address_map[result.subchannel.address()]
    if entry found:
      override_host_attribute.set_actual_address_list(entry.address_list)
  return result
```

### Temporary environment variable protection

This design does not provide any new functionality that will be enabled
by external I/O, so no environment variable guard is necessary.

## Rationale

> We considered implementing nested aggregate clusters as nested cds LB

Author: That's correct.  This flattening is not part of the proposal; it's an alternative that was considered but ruled out for the reasons explained here.

(Most gRFCs don't mention alternatives outside of the "rationale" section.  The dualstack gRFC was an exception to this, since everything was so tightly coupled that it was hard to describe the "what" without the "why", but it was an outlier.)

> We considered implementing nested aggregate clusters as nested cds LB

Reviewer 3: This rationale section mentions two options: tree or flattening. We have to choose one or the other. Neither is mentioned in the proposal. It sounded like you wanted flattening:

> However, since the XdsDependencyManager already needs to resolve the
aggregate cluster dependency graph to make sure the entire config is
loaded, it seemed fairly straightforward to pass down the flattened list
and deal with it all in a single layer.

> We considered implementing nested aggregate clusters as nested cds LB

Author: The flattening isn't mentioned in the proposal, because it's not actually new in this proposal.  We were already doing it that way previously, so we're just sticking with that approach.

I've changed the wording here to try to clarify this.
